File: sundries/OLCodeJudge/HackerRank/tutorials/30-days-of-code/day11.cs
/*
 * Tutorial: 30 Days of Code.
 * A solution to "Day 11: 2D Arrays"
 *   Submitted by A. S. "Aleksey" Ahmann <alekseyyy2@outlook.com>
 *   Submitted on Feb. 24, 2024
 *   Link: https://www.hackerrank.com/challenges/30-2d-arrays/problem
 * 
 * Objective: Today, we are building on our knowledge of arrays by adding another dimension.
 */

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

class Solution {
  
    static void Main(String[] args) {
        int[][] arr = new int[6][];
        for(int arr_i = 0; arr_i < 6; arr_i++){
           string[] arr_temp = Console.ReadLine().Split(' ');
           arr[arr_i] = Array.ConvertAll(arr_temp,Int32.Parse);
        }
        
        // Calculate The Hourglass Sums...
        List<int> hourglass_sums = new List<int>();
        
        for (int row = 1; row < 5; row++) {
            for (int column = 1; column < 5; column++) {
                int x_1 = arr[row][column]; //hourglass centre
                
                int x_2 = arr[row - 1][column]; //hourglass midtop
                int x_3 = arr[row + 1][column]; //hourglass midbottom
                
                int x_4 = arr[row - 1][column - 1]; //hourglass topleft
                int x_5 = arr[row - 1][column + 1]; //hourglass topright
                int x_6 = arr[row + 1][column - 1]; //hourglass bottomleft
                int x_7 = arr[row + 1][column + 1]; //hourglass bottomright
                
                int hourglass_sum = x_1 + x_2 + x_3 + x_4 + x_5 + x_6 + x_7;
                hourglass_sums.Add(hourglass_sum);
            }
        }
        
        // Determine the winner through a voting method...
        int theWinner = hourglass_sums[0];
        for (int i = 1; i < hourglass_sums.Count; i++) {
            if (hourglass_sums[i] > theWinner)
                theWinner = hourglass_sums[i];
        }
        Console.WriteLine("{0}", theWinner);
    }
}

File: sundries/OLCodeJudge/HackerRank/tutorials/30-days-of-code/day11.cpp
/*
 * Tutorial: 30 Days of Code.
 * A solution to "Day 11: 2D Arrays"
 *   Submitted by A. S. "Aleksey" Ahmann <alekseyyy2@outlook.com>
 *   Submitted on Feb. 28, 2024
 *   Link: https://www.hackerrank.com/challenges/30-2d-arrays/problem
 * 
 * Task description: 
 */


#include <bits/stdc++.h>

using namespace std;

int HourglassSum(vector<vector<int>> matrix, int row, int col);
int MaximalHGSum(vector<int> hgSums);

string ltrim(const string &);
string rtrim(const string &);
vector<string> split(const string &);

int main()
{
    vector<vector<int>> arr(6);
    for (int i = 0; i < 6; i++) {
        arr[i].resize(6);

        string arr_row_temp_temp;
        getline(cin, arr_row_temp_temp);

        vector<string> arr_row_temp = split(rtrim(arr_row_temp_temp));

        for (int j = 0; j < 6; j++) {
            int arr_row_item = stoi(arr_row_temp[j]);

            arr[i][j] = arr_row_item;
        }
    }
    
    vector<int> hourglassSums;
    for (int row = 1; row < 5; row++) {
        for (int col = 1; col < 5; col++) {
            int hourglassSum_crc = HourglassSum(arr, row, col);
            hourglassSums.push_back(hourglassSum_crc);
        }
    }
    
    int maxHourglassSum = MaximalHGSum(hourglassSums);
    cout << maxHourglassSum << endl;

    return 0;
}

int HourglassSum(vector<vector<int>> matrix, int row, int col) {
    int e1 = matrix[row][col]; // hour-glass centre
    int e2 = matrix[row - 1][col]; //hour-glass mid-top
    int e3 = matrix[row + 1][col]; //hour-glass mid-bottom
    int e4 = matrix[row - 1][col - 1]; //hour-glass top-left
    int e5 = matrix[row - 1][col + 1]; //hour-glass top-right
    int e6 = matrix[row + 1][col - 1]; //hour-glass bottom-left
    int e7 = matrix[row + 1][col + 1]; //hour-glass bottom-right
    return e1 + e2 + e3 + e4 + e5 + e6 + e7;
}

int MaximalHGSum(vector<int> hgSums) {
    int max = hgSums[0];
    for (int sum : hgSums) {
        if (sum > max)
            max = sum;
    }
    return max;
}

string ltrim(const string &str) {
    string s(str);

    s.erase(
        s.begin(),
        find_if(s.begin(), s.end(), not1(ptr_fun<int, int>(isspace)))
    );

    return s;
}

string rtrim(const string &str) {
    string s(str);

    s.erase(
        find_if(s.rbegin(), s.rend(), not1(ptr_fun<int, int>(isspace))).base(),
        s.end()
    );

    return s;
}

vector<string> split(const string &str) {
    vector<string> tokens;

    string::size_type start = 0;
    string::size_type end = 0;

    while ((end = str.find(" ", start)) != string::npos) {
        tokens.push_back(str.substr(start, end - start));

        start = end + 1;
    }

    tokens.push_back(str.substr(start));

    return tokens;
}

